src/snarks/groth16/mpc/phase2.hpp

```diff
@@ -0,0 +1,309 @@
+#ifndef __ZETH_SNARKS_GROTH16_MPC_PHASE2_HPP__
+#define __ZETH_SNARKS_GROTH16_MPC_PHASE2_HPP__
+
+#include "include_libsnark.hpp"
+#include "snarks/groth16/mpc/hash_utils.hpp"
+
+// Structures and operations related to the "Phase 2" MPC described in
+// [BoweGM17].  Parts of the implementation use techniques from the
+// "Phase2" library from "zk-SNARK MPCs, made easy".
+//
+// References:
+//
+// \[BoweGM17]
+//  "Scalable Multi-party Computation for zk-SNARK Parameters in the Random
+//  Beacon Model"
+//  Sean Bowe and Ariel Gabizon and Ian Miers,
+//  IACR Cryptology ePrint Archive 2017,
+//  <http://eprint.iacr.org/2017/1050>
+//
+// "Phase2" (From "zk-SNARK MPCs, made easy" library
+//  https://github.com/ebfull/phase2)
+//
+// "Sapling MPC" ("Multi-party computation for Zcash's Sapling zk-SNARK public
+//  parameters" https://github.com/zcash-hackworks/sapling-mpc)
+namespace libzeth
+{
+
+template<typename ppT> class srs_powersoftau;
+template<typename ppT> class srs_mpc_layer_L1;
+
+/// Target of the MPC for Phase2 of the SRS generation.  Follows exactly $M_2$
+/// in section 7.3 of [BoweGM17], whre we use L in place of K, consistent with
+/// the keypair in libsnark.
+template<typename ppT> class srs_mpc_phase2_accumulator
+{
+public:
+    // Hash of the initial state before any contributions are made.  Kept
+    // constant over the MPC, and used to check that challenges and responses
+    // are part of the same MPC.
+    srs_mpc_hash_t cs_hash;
+
+    libff::G1<ppT> delta_g1;
+
+    libff::G2<ppT> delta_g2;
+
+    // { H_i } = { [ t(x) . x^i / delta ]_1 }  i=0..n-2 (n-1 entries)
+    libff::G1_vector<ppT> H_g1;
+
+    // { L_j } = { [ ABC_j / delta ]_1 } j=(num_inputs + 1)..num_variables
+    libff::G1_vector<ppT> L_g1;
+
+    srs_mpc_phase2_accumulator(
+        const srs_mpc_hash_t cs_hash,
+        const libff::G1<ppT> &delta_g1,
+        const libff::G2<ppT> &delta_g2,
+        libff::G1_vector<ppT> &&H_g1,
+        libff::G1_vector<ppT> &&L_g1);
+
+    bool operator==(const srs_mpc_phase2_accumulator<ppT> &other) const;
+    bool is_well_formed() const;
+    void write(std::ostream &out) const;
+    void write_compressed(std::ostream &out) const;
+    static srs_mpc_phase2_accumulator<ppT> read(std::istream &in);
+    static srs_mpc_phase2_accumulator<ppT> read_compressed(std::istream &in);
+};
+
+/// Public Key representing a single contribution to the MPC.  This includes a
+/// proof-of-knowledge, consisting of some point $s$ in G1, chosen randomly, $s
+/// * \delta$, $r$ generated deterministically by hashing $s$ and $s * \delta$
+/// into G2, and $r * \delta$.
```

Comment:
The transcript should also show up inside the `hash_to_G2` when computing `r`, right?

Reply:
I've added a mention of the transcript digest.